### PR TITLE
Add logging of tx submission

### PR DESCRIPTION
**Pull request by rvl**
### Issue Number

ADP-630 / ADP-764

### Overview

These are some logging/worker improvements I made in #2570. I'll put them in a separate PR because they overlap with #2583.

- Adds logging of transaction submission results in the wallet layer.
- Fixes the logging types in the tracer bundle.
- Removes duplication between `initWorker` and `registerWorker`.

### CODE REVIEW

File: lib/core/src/Cardano/Wallet/Registry.hs [hunk 1]
```diff
@@ -250,28 +246,50 @@ register registry ctx k (MkWorker before main after acquire) = do
...
             | otherwise -> T.take 8 (toText key) <> ": " <> toText msg
 
 instance HasPrivacyAnnotation (WorkerLog key msg)
 instance HasSeverityAnnotation msg => HasSeverityAnnotation (WorkerLog key msg) where
     getSeverityAnnotation = \case
-        MsgFinished -> Notice
+        MsgThreadAfter msg -> getSeverityAnnotation msg
+        MsgFromWorker _ msg -> getSeverityAnnotation msg
+
+data AfterThreadLog
```
Comment (Anviking) [inline lib/core/src/Cardano/Wallet/Registry.hs:269]: 👍 (short doc comment could be nice though)

File: lib/core/src/Cardano/Wallet/Api/Server.hs [hunk 2]
```diff
@@ -2370,26 +2321,74 @@ registerWorker
...
+  where
+    before ctx' _ = void $ unsafeRunExceptT $ createWallet ctx'
+    re = ctx ^. workerRegistry @s @k
+
+-- | Something to pass as the coworker action to 'newApiLayer', which does
+-- nothing, and never exits.
+idleWorker :: ctx -> wid -> IO ()
+idleWorker _ _ = forever $ threadDelay maxBound
+
+startWorker
```
Comment (Anviking) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]: Some short doc-comments connecting startWorker, registerWorker and initWorker would be useful

File: lib/core/src/Cardano/Wallet/Api/Server.hs [hunk 3]
```diff
@@ -3009,3 +3008,29 @@ instance LiftHandler (Request, ServerError) where
...
                 ) : headers
             }
+
+{-------------------------------------------------------------------------------
+                               Logging
+-------------------------------------------------------------------------------}
+
+-- | The type of log messages coming from the server 'ApiLayer', which may or
+-- may not be associated with a particular worker thread.
+data WalletEngineLog
```
Comment (Anviking) [inline lib/core/src/Cardano/Wallet/Api/Server.hs:3017]: 👌
But you have `MsgTxSubmit` both in `WalletEngineLog` _and_ in `WalletLog`? 🤔

Is this intentional? For txs when first submitted from a wallet, and for when they are retried?

File: lib/core/src/Cardano/Wallet.hs [hunk 4]
```diff
@@ -1557,38 +1562,45 @@ submitTx
...
+        , HasLogger WalletLog ctx
         )
     => ctx
     -> WalletId
     -> (Tx, TxMeta, SealedTx)
     -> ExceptT ErrSubmitTx IO ()
 submitTx ctx wid (tx, meta, binary) = db & \DBLayer{..} -> do
-    withExceptT ErrSubmitTxNetwork  $
+    withExceptT ErrSubmitTxNetwork $ traceWithExceptT tr $
         postTx nw binary
```
Comment (Anviking) [inline lib/core/src/Cardano/Wallet.hs:1573]: I think tracing here makes sense, but `postTx` also has a debug-level `MsgPostTx`, which is arguably a bit weird

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs [hunk 5]
```diff
@@ -896,10 +883,7 @@ instance HasSeverityAnnotation StakePoolLog where
         MsgFetchPoolMetadata e -> getSeverityAnnotation e
         MsgFetchTakeBreak{} -> Debug
         MsgGCTakeBreak{} -> Debug
-        MsgGCFinished{} -> Debug
-        MsgGCThreadKilled{} -> Debug
-        MsgGCUserInterrupt{} -> Debug
-        MsgGCUnhandledException{} -> Debug
```
Comment (Anviking) [inline lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:902]: nice

**Review by Anviking — APPROVED** [review]
👌 think it makes sense

File: lib/core/src/Cardano/Wallet/Api/Server.hs (hunk 3 above)
Comment (rvl) [reply lib/core/src/Cardano/Wallet/Api/Server.hs:3017]: Yes - one is for the raw transaction submit endpoint - which is not associated with any wallet.
I'll rename the constructor though, to make this less confusing.

File: lib/core/src/Cardano/Wallet.hs (hunk 4 above)
Comment (rvl) [reply lib/core/src/Cardano/Wallet.hs:1573]: Yes - that logging is good for debug, but it's too low-level, and doesn't log the result.
The transaction submission event is something that the user should be told about with nice log messages.

File: lib/core/src/Cardano/Wallet/Registry.hs (hunk 1 above)
Comment (rvl) [reply lib/core/src/Cardano/Wallet/Registry.hs:269]: OK done.

File: lib/core/src/Cardano/Wallet/Api/Server.hs (hunk 2 above)
Comment (rvl) [reply lib/core/src/Cardano/Wallet/Api/Server.hs]: Added a doc comment, and renamed these functions to be more descriptive.

**Comment by rvl** [comment]
Thanks for the review comments, they should be addressed. Let's try our luck...

bors r+

**Pull request merged**